packages/backend-common/src/service/lib/ServiceBuilderImpl.ts

```diff
@@ -121,20 +145,91 @@ export class ServiceBuilderImpl implements ServiceBuilder {
         reject(e);
       });
 
-      const server = stoppable(
-        app.listen(port, host, () => {
-          logger.info(`Listening on ${host}:${port}`);
-        }),
-        0,
-      );
+      let server: http.Server;
+
+      if (httpsSettings) {
+        logger.info('Initializing https server');
+
+        const credentials: { key: string; cert: string } = {
+          key: '',
+          cert: '',
+        };
+        const signingOptions: any = httpsSettings?.certificate;
+
+        if (signingOptions?.algorithm !== undefined) {
+          logger.info('Generating self-signed certificate with attributes');
+
+          const certificateAttributes: Array<any> = Object.entries(
+            signingOptions.attributes,
+          ).map(([name, value]) => ({ name, value }));
+
+          // TODO: Create a type def for selfsigned.
+          const signatures = require('selfsigned').generate(
+            certificateAttributes,
+            {
+              algorithm: signingOptions?.algorithm,
+              keySize: signingOptions?.size || 2048,
+              days: signingOptions?.days || 30,
+            },
+          );
+
+          logger.info(
+            'Bootstrapping key and cert from self-signed certificate',
+          );
+
+          credentials.key = signatures.private;
+          credentials.cert = signatures.cert;
+        } else {
+          if (fs.existsSync(signingOptions?.key)) {
```

Comment:
hm, could split into `key` and `keyFile`, `cert` and `certFile` perhaps ... that would make it easier to treat the errors too.

perhaps even three possibilities:

```yaml
certificate:
  # ... either ...
  file:
    keyPath: ./blah
    certPath: ./blah
  # ... or ...
  raw:
    key: lsdfsdlfdsl
    cert: dslfksldfnsdlfn 
  # ... or ...
  generated:
    algorithm: lfksjflsd
```

that would lead to cleaner types and responsibilities in the code

Reply:
I already had a length debate with Patrick about this in the previous PR. As such I am not going to spent to much time on extending this logic as the PR cycles are tearing into our velocity (big time) and these types of decisions are better handled by the core team. As it stands we can already support the three scenarios, two of them are managed by the CertificateReferenceOptions, which obviously violates the Single Responsibility Principle however since we do not see eye to eye on the "root dir" issue (Patrick wants it to resolve to the root of the repo, I want it to resolve to the root of the node module) I figured it was better to let you guys fit it into your architecture after you have accepted my PR into core and let me focus on our own sprint deliveries so our PO will fund us to work more on this project ;)

Reply:
I do not want to resole relative to the root of the repo. I want to use the built in secrets management of the config system to read in these files, which has the option to read secrets from files relative to each app-config.yaml

Reply:
Oh of course, that makes sense. I forgot that that applies nicely here. @ZaradarDFDS in order to get you through this quickly and able to merge, would you think it OK to actually REMOVE the file support and then be done? The rest is basically nitpicking and I'd be happy to get it in soon and unblock you and iterate

Reply:
I think maybe I am confusing things here a bit due to my limited knowledge of Backstage (compared to you guys), so for the sake of clarity I will try to rephrase myself. As such we have three different ways to attach a certificate to our express server:

1) Using a auto-generated-self-signed-certificate (this is what .NET core does and what I prefer)
2) Using a certificate file. This is typically what you would get in a legacy enterprise.
3) Using a k8s secret (or some other secret) which is what we have at Spotify and DFDS.

Our "service factory" needs to cater to all the above scenarios. This means we should be able to support the three configurations that freben suggested (file, raw (== secret) and generated (== self-signed)). Now as I see it the "file" scenario has very little to do with secrets and everything to do with how our Node JS process will resolve the path that is provided to our code. Seeing how the backend (and all the other standAlone servers) are logically separate processes it makes no sense to redirect relative file paths to the root of our repo, but rather to resolve it to the root of the Node JS process (standalone server), which is the point I am trying to make.

I might be misunderstanding Patricks input, but based on the communication around this issue I have a very hard time seeing how any cert file path provided to our config (regardless of where it comes from) has anything to do with the secret provider?

Example configs:

https (self-signed):
   certificate:
     size: 2048
     algorithm: sha256
     days: 30
   attributes: 
     commonName: 'dfds.com'

https (secret):
   key: 'RAW_CERT_STRING_FROM_SECRET'

https (raw - atm we only support absolute paths until we agree on the relative resolution issue):
   key: 'ABSOLUTE_OR_RELATIVE_PATH_TO_FILE' 


On another note it feels very much like work you could do ALOT more effectively one your own, based on my PR, with the insights you have. It will be almost impossible for any contributor to predict the style of code you prefer when these rules seem highly subjective and are not enforced by lint'ers. Combining this with the Alpha-stage we are in I fear that the cost to our team velocity will get very high if we have to pass some "dynamic-acceptance-criteria" that is not automated and forces scope creep on the contributors backlog because he/she only really cares about the logic inluded in the PR (in our case its the self-signing https stuff) which makes it super hard to justify to a PO that the team will spend a few extra days work satisfying change requests that does not provide their business with any value what so ever. :)

Reply:
I will drop the support for files. We dont need it, my m8s at Ørsted probably will if I convince them to jump on the bandwagon, but at the point in time we can re-implement the feature. :)

Reply:
@ZaradarDFDS I think what @Rugvip meant was to use [this](https://github.com/spotify/backstage/blob/master/docs/conf/writing.md#file-secrets) facility to have the framework do the file loading for you. Then you would only need the string versions of `key` and `cert`, because they would just be populated with the string contents of those files as if you had written them in the yaml. There may be some misunderstanding in relation to how the path resolution works, I'll leave that aside.

I fully appreciate that you are not at liberty to poke at this indefinitely, in particular for things that you do not need yourselves. I am eager to get this in and unblocking you as fast as possible, and we are happy to add any additional tweaks on top of it after that.

Reply:
I appreciate both of your efforts to help and when we are in a more "stable" place with regards to our project team it will also be easier for us to spend more time working with and talking about the overall architecture. However right now we just need a few "quick wins" to show people that this can be a great enabler for collaborating around a "shared" self-service portal. :)